sdks/core/src/main/java/org/gbif/pipelines/core/parsers/dynamic/DynamicParser.java

```diff
@@ -0,0 +1,119 @@
+package org.gbif.pipelines.core.parsers.dynamic;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import lombok.Builder;
+
+@Builder
+class DynamicParser {
+
+  private final Map<String, String> mainTemplateMap = new LinkedHashMap<>();
+  private final List<PatternDefault> patterns = new ArrayList<>();
+  private final Pattern unitsFromKey;
+  private final Map<String, String> keyMap;
+  private final String field;
+
+  protected void addTemplate(String key, String value) {
+    String result = value;
+    for (Map.Entry<String, String> entry : RegexDefaultTemplates.get().entrySet()) {
+      result = result.replace(entry.getKey(), entry.getValue());
+    }
+    mainTemplateMap.put(key, result);
+  }
+
+  protected void initPattern(String value) {
+    initPattern(value, null, null);
+  }
+
+  protected void initPattern(String value, String defaultKey, String defaultType) {
+    String result = value;
+    for (Map.Entry<String, String> entry : mainTemplateMap.entrySet()) {
+      result = result.replace(entry.getKey(), entry.getValue());
+    }
+    for (Map.Entry<String, String> entry : RegexDefaultTemplates.get().entrySet()) {
+      result = result.replace(entry.getKey(), entry.getValue());
+    }
+    patterns.add(
+        PatternDefault.builder()
+            .pattern(Pattern.compile(result))
+            .defaultKey(defaultKey)
+            .defaulType(defaultType)
+            .build());
+  }
+
+  protected String getGroup(Matcher matcher, String value, String... group) {
+    String result = null;
+    for (String g : group) {
+      if (matcher.pattern().pattern().contains("<" + g + ">")) {
+        String s = matcher.group(g);
+        if (result == null) {
+          result = s;
+        } else {
+          result = String.join(", ", result, s);
+        }
+      }
+    }
+    return result == null ? value : result;
+  }
+
+  protected void initTemplates() {
```

Comment:
not really sure what is happening here...it is keeping the last value of a multimap?...not sure if that should be called init?